x/iro/keeper/settle.go

```diff
@@ -0,0 +1,151 @@
+	futBalance := k.BK.GetBalance(ctx, k.AK.GetModuleAddress(types.ModuleName), plan.TotalAllocation.Denom)
+	err := k.BK.BurnCoins(ctx, types.ModuleName, sdk.NewCoins(futBalance))
+	if err != nil {
+		return err
+	}
+
+	// mark the plan as `settled`, allowing users to claim tokens
+	plan.SettledDenom = rollappIBCDenom
+	k.SetPlan(ctx, plan)
+
+	// uses the raised DYM and unsold tokens to bootstrap the rollapp's liquidity pool
+	err = k.bootstrapLiquidityPool(ctx, plan)
+	if err != nil {
+		return errors.Join(types.ErrFailedBootstrapLiquidityPool, err)
+	}
+
+	// Emit event
+	err = ctx.EventManager().EmitTypedEvent(&types.EventSettle{
+		RollappId: rollappId,
+		PlanId:    fmt.Sprintf("%d", plan.Id),
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Bootstrap liquidity pool with the raised DYM and unsold tokens
+func (k Keeper) bootstrapLiquidityPool(ctx sdk.Context, plan types.Plan) error {
+	unallocatedTokens := plan.TotalAllocation.Amount.Sub(plan.SoldAmt)        // assumed > 0, as we enforce it in the Buy function
+	raisedDYM := k.BK.GetBalance(ctx, plan.GetAddress(), appparams.BaseDenom) // assumed > 0, as we enforce it by IRO creation fee
+
+	// send the raised DYM to the iro module
+	err := k.BK.SendCoinsFromAccountToModule(ctx, plan.GetAddress(), types.ModuleName, sdk.NewCoins(raisedDYM))
+	if err != nil {
+		return err
+	}
+
+	// calculate last price
+	lastPrice := plan.BondingCurve.SpotPrice(plan.SoldAmt)
+	// find the tokens needed to bootstrap the pool, to fulfill last price
+	tokens, dym := determineLimitingFactor(unallocatedTokens, raisedDYM.Amount, lastPrice)
+	rollappLiquidityCoin := sdk.NewCoin(plan.SettledDenom, tokens)
+	dymLiquidityCoin := sdk.NewCoin(appparams.BaseDenom, dym)
+
+	// create pool
+	gammGlobalParams := k.gk.GetParams(ctx).GlobalFees
+	poolParams := balancer.NewPoolParams(gammGlobalParams.SwapFee, gammGlobalParams.ExitFee, nil)
+	balancerPool := balancer.NewMsgCreateBalancerPool(k.AK.GetModuleAddress(types.ModuleName), poolParams, []balancer.PoolAsset{
+		{
+			Token:  dymLiquidityCoin,
+			Weight: math.OneInt(),
+		},
+		{
+			Token:  rollappLiquidityCoin,
+			Weight: math.OneInt(),
+		},
+	}, "")
+
+	// we call the pool manager directly, instead of the gamm keeper, to avoid the pool creation fee
+	poolId, err := k.pm.CreatePool(ctx, balancerPool)
+	if err != nil {
+		return err
+	}
+
+	// Add incentives
+	poolDenom := gammtypes.GetPoolShareDenom(poolId)
+	incentives := sdk.NewCoins(
+		sdk.NewCoin(dymLiquidityCoin.Denom, raisedDYM.Amount.Sub(dymLiquidityCoin.Amount)),
+		sdk.NewCoin(rollappLiquidityCoin.Denom, unallocatedTokens.Sub(rollappLiquidityCoin.Amount)),
+	)
+	distrTo := lockuptypes.QueryCondition{
+		LockQueryType: lockuptypes.ByDuration,
+		Denom:         poolDenom,
+		Duration:      k.ik.GetLockableDurations(ctx)[0],
+	}
+	_, err = k.ik.CreateGauge(ctx, false, k.AK.GetModuleAddress(types.ModuleName), incentives, distrTo, ctx.BlockTime().Add(time.Hour), 30)
```

Comment:
the 30 epochs should prob be a gov param.

Reply:
my bad. it's set by the plan creator